### DISCUSSION TITLE
ERCF multi material system with the 5M

### DISCUSSION THREAD

**Discussion by peterxn123** [xblax/flashforge_ad5m_klipper_mod#181] [Q&A]
Hi all,

Firstly, I wanted to thank the people that made this mod possible as it is truly amazing and works a treat. This is also my first time writing anything in a discussion on GitHub, so please excuse me if I've done anything wrong.

I was recently browsing aliexpress when I saw a kit for an Enraged Rabbit Carrot Feeder, a multi material system originally for the Voron 2.4 but also compatible with all other klipper printers. [This](https://github.com/EtteGit/EnragedRabbitProject/tree/main) is the GitHub for the project. The documentation is [here](https://github.com/EtteGit/EnragedRabbitProject/tree/main/Documentation).

I've been wanting to install a multi material system on my 5M for a while now and have been patiently waiting for a release from Flashforge. However, I am now left here wondering whether or not it would be possible to do this mod with my newly klipper modded 5M.

Could someone cleverer than me take a look and see if this would be possible to add to a klipper modded 5M please? It doesn't actually require any hardware mods for the printer, just seems to be some klipper configs and macros. The bit that I am unsure about is that it requires a separate controller board to control the multi material system, which it says should then be connected to the klipper system.

If it is possible, I will go and buy one ASAP and report back to you.

Thanks.

**Comment by Avpman2** [comment]
+1 for this!

**Comment by KaruroChori** [comment]
Should it not support a filament cutter for best compatibility?

**Comment by consp** [comment]
>  It doesn't actually require any hardware mods for the printer

You might want to add a USB port as it would use up the only one available to connect the helper board, but there are plenty available on the board, just not with connectors attached.

**Reply by peterxn123** [reply]
Yeah, I could do that. Other than that, do you think it would be a possible mod for the 5M? Because otherwise if I bought it and couldn't get it working it'd become a paperweight until I got another klipper printer.

**Reply by consp** [reply]
I have no clue, never used it on any printer and just skimmed the manual quickly, and as long as it's compatible with klipper v0.11 it might work, but I have no clue about the actual requirements. If you know those it might be a bit easier to check it out. I don't think it's too complicated but someone with more klipper experience needs to weigh in on this.

**Comment by Avpman2** [comment]
Has anyone tried to hook this into the mod? I looked at two other feeder projects, and the ERCF, while not the most simplistic, seems the most viable option. I want to give it a try if anyone else is game.

**Reply by KaruroChori** [reply]
Slowly working on it. The biggest issue is the placement of a filament cutter.

**Reply by Avpman2** [reply]
What about software/FW configuration with this Klipper mod? Do you have it running?

**Reply by KaruroChori** [reply]
Not really since I am still assembling it. They are just normal klipper macros, there is no reason why they should not work out of the box.

**Reply by Avpman2** [reply]
Thanks, anxious to hear how it goes for you. BTW - how does it connect hardware-wise to the printer (USB, etc)?

**Reply by KaruroChori** [reply]
There is a usb port on the display unit, I am using that one.
There is also a second header which was designed for the camera, that would work as well with some crimping and soldering if keeping the usb for storage was a priority of mine.
As for power, I am using an external power supply.

In an ideal world it would be a second canbus device, but that would be a bit hard to setup on this printer.

**Reply by function3d** [reply]
> There is a usb port on the display unit, I am using that one

Could you tell me what is the system path to the usb port on the display or how can I find out? Thanks

**Reply by consp** [reply]
Might also need a mdev rule @KaruroChori got any idea if it will auto-detect or not?

**Comment by skipster74** [comment]
I’ve also been fixated with multifiliment options. I was looking at this:

https://www.3dchameleon.com/

it seems a hare less complicated than the rabbit.

**Reply by Avpman2** [reply]
I saw that one. I'm concerned about the mechanical triggering required to switch filaments. The Rabbit is triggered electronically. The Chameleon requires the bed or something to make contact with a switch x times to change filaments.

**Comment by f00d4tehg0dz** [comment]
@Avpman2  I have a near-complete functional ERCF v1.1 with ERCF Filament Cutter mod running on the AD5M Pro. (I am dealing with some unloading issues that I'm trying to debug)

This is the commit I got working to install [Happy-Hare](https://github.com/moggieuk/Happy-Hare/tree/5d84cb25b83c494f0224929000f45e2a46678e21)  

I also needed to recompile the AD5M mod to allow for `mkdir -p /dev/serial/by-id` and `sudo`.
Then you can add your USB EasyBRD. 

If you want to skip compiling the mod yourself. You can grab the recompiled [pkg ](https://github.com/f00d4tehg0dz/Flashforge-5M-5M-Pro---Happy-Hare/tree/flashforge-klipper-packages)

An example config is [there too](https://github.com/f00d4tehg0dz/Flashforge-5M-5M-Pro---Happy-Hare/tree/flashforge-config)

Disregard this part. It appears moggieuk has updated to run on Klipper 11 as of yesterday. Here is the commit to copy into your Happy-Hare. https://github.com/moggieuk/Happy-Hare/pull/388/commits
~~Happy-Hare with the cutter mod and a few QOL from the latest Happy Hare merged is included [here](https://github.com/f00d4tehg0dz/Flashforge-5M-5M-Pro---Happy-Hare/tree/dec-8-2023-working-flashforge):~~

To get your USB controller working. It's quite easy: 

`mkdir -p /dev/serial/by-id/`

`sudo ln -s /dev/ttyACM0 /dev/serial/by-id/usb-Klipper_samd21g18a_xxx-if00`

Verify that the connection for ttyACM0 (or whichever number for you) is now purple
`ls /dev/serial/by-id/*`

Create a file named `S99klipper_serial` with the following content:
```#!/bin/sh

# Create the directory if it doesn't exist
mkdir -p /dev/serial/by-id/

# Create the symbolic link
ln -sf /dev/ttyACM0 /dev/serial/by-id/usb-Klipper_samd21g18a_xxx-if00
`ln -sf /dev/ttyACM0

# List the contents of the directory for verification
ls /dev/serial/by-id/*```

Ensure the script is executable:
`chmod +x /etc/init.d/S99klipper_serial`


git clone https://github.com/moggieuk/Happy-Hare.git
cd /Happy-Hare
~~git checkout 7915199748f38b8564ec970c5aa7b510cdffa914~~

Replace check_klipper function with this
```check_klipper() {
    KLIPPER_PID_FILE="/run/klipper.pid"
    if [ "$NOSERVICE" -ne 1 ]; then
        if [ -f "$KLIPPER_PID_FILE" ]; then
            echo -e "${INFO}Klipper service running (PID file found: $KLIPPER_PID_FILE)"
        else
            echo -e "${ERROR}Klipper service not found! Please install Klipper first"
            exit -1
        fi
    fi
}
```

And comment out verify_not_root logic.
Lastly, comment out exit -1 in verify_not_root
```verify_not_root() {
    if [ "$EUID" -eq 0 ]; then
        echo -e "${ERROR}This script must not run as root"
        #exit -1
    fi
}
```
install.sh with -i 
`./install.sh -k /root/printer_software/klipper -m /root/printer_software/moonraker/ -i`
```

Edit your moonraker cfg in /root/printer_data/config
```[machine]
provider: systemd_cli
```

**Reply by KaruroChori** [reply]
That is great! Which mod are you using for the cutter? I tried to add one to the toolhead, but I had issues automatically triggering it with most materials.

**Reply by f00d4tehg0dz** [reply]
> That is great! Which mod are you using for the cutter? I tried to add one to the toolhead, but I had issues automatically triggering it with most materials.

This is the one. [Version 7 in Doc and STL's(beta)](https://github.com/kevinakasam/ERCF_Filament_Cutter)

I'm using this [5v step-down converter.](https://www.amazon.com/dp/B07JZ2GQJF?ref=ppx_yo2ov_dt_b_fed_asin_title ) 
It's plugged into the 12v that powers the servo in EasyBRD.

I went through so many variations of filament cutters till I got one working. haha.

**Comment by f00d4tehg0dz** [comment]
I also made this attachment to mount to keep the Bowden from popping out on loading filament. 
![image](https://github.com/user-attachments/assets/1bdf0e3a-ddf9-4d25-ad24-0ff21f51cb2b)
OBJ included in the zip below
[Flashforge 5DM Pro Extruder Fitting.zip](https://github.com/user-attachments/files/17019744/Flashforge.5DM.Pro.Extruder.Fitting.zip)

**Comment by wwirbel** [comment]
Hi, I am tring to rebuild the ercf config with my 5m non pro and an ercf 1.1 with usb-Klipper_rp2040 (ERB) , but I am unable to connect to the MCU (the board was preflashed on a different system and was working there).
Sobody has an Idea how I can continue tu integrate it?
I tried the configs from [f00d4tehg0dz] and also configured it out of the box, without succes.

If I connect the board it is detecte as ttyACM0 
so I used 
 ln -sf /dev/ttyACM0 /dev/serial/by-id/usb-Klipper_rp2040_E6612C771F7
D742B-if0
and used in mmu.cfg
[mcu mmu]
setups
serial: /dev/serial/by-id/usb-Klipper_rp2040_E6612C771F7D742B-if00	

I didnt find (make menuconfig) so, I still used the preflashed board .

Did someone has a hint for me to make it work, many thanks in adavance.

**Reply by wwirbel** [reply]
[klipper(4).log](https://github.com/user-attachments/files/18262713/klipper.4.log)

after a complet reinstall I got : 

    cb()
  File "/root/printer_software/klipper/klippy/mcu.py", line 73, in _build_config
    self._trdispatch_mcu = ffi_main.gc(ffi_lib.trdispatch_mcu_alloc(
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
OverflowError: can't convert negative number to unsigned
Internal error during connect: can't convert negative number to unsigned

Once the underlying issue is corrected, use the "RESTART"
command to reload the config and restart the host software.
Printer is halted

Script running error
Traceback (most recent call last):
  File "/root/printer_software/klipper/klippy/extras/gcode_button.py", line 41, in button_callback
    self.gcode.run_script(template.render())
  File "/root/printer_software/klipper/klippy/gcode.py", line 216, in run_script
    self._process_commands(script.split('\n'), need_ack=False)
  File "/root/printer_software/klipper/klippy/gcode.py", line 198, in _process_commands
    handler(gcmd)
  File "/root/printer_software/klipper/klippy/gcode.py", line 272, in cmd_default
    raise gcmd.error(self.printer.get_state_message()[0])
gcode.CommandError: Internal error during connect: can't convert negative number to unsigned

**Reply by f00d4tehg0dz** [reply]
Try one of the precompiled packages for now. I clearly need to update my instructions as something is wrong here for you. 
https://github.com/f00d4tehg0dz/Flashforge-5M-5M-Pro---Happy-Hare/tree/flashforge-klipper-packages

**Reply by Plastic-armyman** [reply]
> Hi, I am tring to rebuild the ercf config with my 5m non pro and an ercf 1.1 with usb-Klipper_rp2040 (ERB) , but I am unable to connect to the MCU (the board was preflashed on a different system and was working there). Sobody has an Idea how I can continue tu integrate it? I tried the configs from [f00d4tehg0dz] and also configured it out of the box, without succes.
> 
> If I connect the board it is detecte as ttyACM0 so I used ln -sf /dev/ttyACM0 /dev/serial/by-id/usb-Klipper_rp2040_E6612C771F7 D742B-if0 and used in mmu.cfg [mcu mmu] setups serial: /dev/serial/by-id/usb-Klipper_rp2040_E6612C771F7D742B-if00
> 
> I didnt find (make menuconfig) so, I still used the preflashed board .
> 
> Did someone has a hint for me to make it work, many thanks in adavance.

Ok, sorry to Necro this.

If I am correct, using this you were able to get Serial over USB on stock firmware on the AD5M?

I'm trying to get Serial over USB working on my AD5X so I can use Octoprint, I'm thinking this might be the right route

**Comment by wwirbel** [comment]
Install mod from:
https://github.com/f00d4tehg0dz/Flashforge-5M-5M-Pro---Happy-Hare/blob/flashforge-klipper-packages/Adventurer5M-KlipperMod-v00.05-beta-5-gdc421d0-klipperscreen.tgz   for non Pro
mkdir -p /dev/serial/by-id/
ln -sf /dev/ttyACM0 /dev/serial/by-id/usb-Klipper_rp2040_E6612C771F7D742B-if00  (thats the id from my old printers config
git clone https://github.com/moggieuk/Happy-Hare.git
git checkout 7915199748f38b8564ec970c5aa7b510cdffa914
edit install.sh   / verify not root and check klipper
./install.sh -k /root/printer_software/klipper -m /root/printer_software/moonraker/ -i

1 for ERCF V1.1, no options    , 2 for Borrow ERB   6 Gates rest no expept include
add Moonracker config : provider: systemd_cli

same issue durring connect

Sending MCU 'mcu' printer configuration...
Configured MCU 'mcu' (1024 moves)
Unhandled exception during connect
Traceback (most recent call last):
  File "/root/printer_software/klipper/klippy/klippy.py", line 180, in _connect
    cb()
  File "/root/printer_software/klipper/klippy/mcu.py", line 732, in _connect
    self._send_config(None)
  File "/root/printer_software/klipper/klippy/mcu.py", line 662, in _send_config
    cb()
  File "/root/printer_software/klipper/klippy/mcu.py", line 73, in _build_config
    self._trdispatch_mcu = ffi_main.gc(ffi_lib.trdispatch_mcu_alloc(
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
OverflowError: can't convert negative number to unsigned
Internal error during connect: can't convert negative number to unsigned

**Reply by f00d4tehg0dz** [reply]
I see. Remove the `git checkout 7915199748f38b8564ec970c5aa7b510cdffa914` that is not needed now. Happy-Hare has been updated recently. 
I would remove the repo folder and git clone again.

**Comment by wwirbel** [comment]
Same result, 
for testing I just add the mcu configuration to the printer.cfg, that was working for a few secons then I lost conctact to MMU MCU and the Serial IP Changed from TTYACM0 to TTYACM1.
I thing the board doesnt lice my ERFC ERB V1.1 and since make menuconfig ins not implemented I canont flas the boar with the kliper version.
Thanks for your effort I really appreciated.

**Reply by f00d4tehg0dz** [reply]
Try these more up to date instructions. https://www.reddit.com/r/FlashForge/comments/1gl8bgd/partial_success_for_multiple_filament_ercf_v2/
